man/osprep.1
.TH OSPREP 1
.SH NAME
osprep \- Operating System Preparation Utility
.SH SYNOPSIS
.B osprep
[
\fBupdate\fR [\fIURL\fR] |
\fBbases\fR |
\fBbase\fR [\fIname\fR[~\fIversion\fR]] |
\fBpackages\fR |
\fBadd\fR [\fIname\fR[~\fIversion\fR]] |
\fBremove\fR \fIname\fR[~\fIversion\fR] |
\fBlist\fR |
\fBsnap\fR |
\fBset\fR [\fIname\fR~\fIversion\fR[::\fIkey\fR [\fIvalue\fR]]] |
\fBapply\fR |
\fBdiscard\fR |
\fBimage\fR \fIfile\fR |
{\fB\-v\fR, \fB\-\-version\fR} |
{\fB\-h\fR, \fB\-\-help\fR, \fBhelp\fR}
[{\fB\-l\fR, \fB\-\-local\fR}]
[\fB\-\-format=\fIoutput_format\fR\fR]
]
.SH DESCRIPTION
.B Operating System Preparation Utility
creates a unified workflow for preparing and customizing operating system installation media. The major foundation of this project is to make it target platform independent, that means to be able to process scripts for generating any kind of media of any kind of operating system. It makes it possible for the user to change settings, add packages to the image and to create a ready-to-go disk image.

When run with no parameters, \fBosprep\fR launches an interactive dialog(1) based interface.

When any parameters are given, one of the command below are run.
.TP
.BR update\ [\fIURL\fR]
Update local repository, optionally change remote repository address.
Default \fIURL\fR is http://charlotte.celones.pl/osprep/.
.TP
.BR bases
Show a list of available project base packages.
.TP
.BR base\ [\fIname\fR[~\fIversion\fR]]
Select base package for the project.
Show currently selected base package when no name is given.
Select the newest version when no \fIversion\fR is given.
.TP
.BR packages
Show a list of packages available for the project.
.TP
.BR add\ [\fIname\fR[~\fIversion\fR]]
Add a package to the project.
Select the newest version when no \fIversion\fR is given.
.TP
.BR remove\ \fIname\fR[~\fIversion\fR]
Remove package(s) from the project.
Select all versions when no \fIversion\fR is given.
.TP
.BR list
Show a list of packages added to the project.
.TP
.BR snap
Download and extract selected packages.
.TP
.BR set\ [\fIname\fR~\fIversion\fR[::\fIkey\fR [\fIvalue\fR]]]
Show or modify configuration variables.
Show all variables with their values when no \fIname\fR and \fIversion\fR is given.
Show all variables of a particular package when no \fIkey\fR is given.
Show only value of a variable when no new \fIvalue\fR is given.
.TP
.BR apply
Apply packages and configuration.
.TP
.BR discard
Discard all changes from base.
.TP
.BR image\ \fIfile\fR
Create output disk image.
.TP
.BR \-v,\ \-\-version
Printout version information and exit.
.TP
.BR \-h,\ \-\-help,\ help
Show the help text and exit.
.SH OPTIONS
.TP
.BR \-l,\ \-\-local
Disable checking for package updates
.TP
.BR \-\-format=\fIoutput_format\fR
Set output format.
Supported output formats are:
\fBtext\fR - user-friendly formatted text.
\fBcsv\fR - easy-to-process semicolon-separated dataset.
.SH EXIT STATUS
\fBosprep\fR return a non-zero value on error.
.SH FILES
\fBosprep\fR stores its packagelists and local package repository in \fB.osprep\fR subdirectory of the current user's home directory.

All project settings are stored in \fB.osprep\fR subdirectory of the current working directory.
.SH NOTES
\fBosprep\fR needs bash(1), wget(1) and unxz(1) to run. In order to mount and create disk images the existence of sudo(8) is inevitable. If the user wants to run it with an interactive menu, dialog(1) is going to be necessary as well.
.SH AUTHORS
Copyright © 2018 Mateusz Karcz

This document as a part of \fBosprep\fR is licensed under the MIT License.

GitHub repository: https://github.com/thecatkitty/osprep


4_osprep/osprep.sh
#!/bin/bash -l

# Author      : Mateusz Karcz
# Last edited : May 2018
# Version     : 0.1.1805
#
# Description : A tool for downloading and configuring boot disks of hipster
#               operating systems.
#
# Licensed under the MIT License.

PROGRAM_NAME="Celones Operating System Preparation Utility"
PROGRAM_VERSION="0.1.1805"

HTTP_USER_AGENT="Celones-OSPrep/$PROGRAM_VERSION"

NC="\033[0m"
BOLD="\033[1m"
GREEN="\033[0;32m"

OPT_FORMAT="text"
OPT_LOCAL=false
REPO_URL="http://pkg.svc.celones.pl/osprep/"

############################### HELPER ROUTINES ################################
function check_wget {
  if [ `command -v wget` -eq "" ]; then
    echo "This command requires wget, but it hasn't been found!" >&2
    exit 1
  fi
}

function check_make {
  if [ `command -v make` -eq "" ]; then
    echo "This command requires GNU Make, but it hasn't been found!" >&2
    exit 1
  fi
}

function http_header {
  # $URL    - remote resource
  # $HEADER - name of the HTTP header
  wget -S --spider "$URL" 2>&1 | grep "$HEADER: " | sed -e 's/^ *[^:]\+: //'
}

function http_download {
  # $URL  - remote resource
  # $FILE - output file name
  wget --no-verbose --show-progress --progress=bar -U "$HTTP_USER_AGENT" -O "$FILE" "$URL"
}

function download_gauge {
  # $URL  - remote resource
  # $FILE - output file name
  wget --progress=dot -U "$HTTP_USER_AGENT" -O "$FILE" "$URL" 2>&1 \
  | grep "%" | sed -u -e 's/^ *[^ ]\+[\. ]\+//g' | sed -u -e 's/%.*//' \
  | dialog \
  --backtitle "$PROGRAM_NAME" \
  --title "Downloading" \
  --gauge "$URL" 10 100
}

################################### COMMANDS ###################################
function interactive {
  tput smcup

  RESULT=`dialog \
  --backtitle "$PROGRAM_NAME" \
  --title "Build List Box" \
  --visit-items \
  --buildlist "Text" 0 0 0 \
    "it1" "Initially selected"     on \
    "it2" "Initially not selected" off 2>&1 1>&3`

  dialog \
  --backtitle "$PROGRAM_NAME" \
  --title "Message Box" \
  --msgbox "You've selected: $RESULT" 0 0

  tput rmcup

}

function version {
  echo $PROGRAM_NAME
  echo "Version $PROGRAM_VERSION"
  echo "© 2018 Mateusz Karcz. All rights reserved."
  echo "Licensed under the MIT License."
}

function help {
  echo "Program help"
}

function update {
  FILE=".osprep/cache/bases.csv"
  URL="${REPO_URL}bases.csv"

  mkdir -p $(dirname "$FILE")

  if [ -f "$FILE" ]; then
    HTTP_LAST_MODIFIED=$(HEADER=Last-Modified http_header)
    if [ $(date -d "$HTTP_LAST_MODIFIED" +%s) -gt $(stat -c %Y "$FILE") ]; then
      rm "$FILE"
    else
      echo "Repository already up to date."
    fi
  fi

  if [ ! -f "$FILE" ]; then
    echo "Updating the repository..."
    http_download
    if [ -f "$FILE" ]; then
      echo "Repository update completed."
    else
      echo "Cannot update repository!" >&2
      exit 1
    fi
  fi
}

function bases {
  FILE=".osprep/cache/bases.csv"

  if [ ! -f "$FILE" ]; then update; fi

  if [ $OPT_FORMAT == "csv" ]; then cat "$FILE"

  else
    cat "$FILE" | while read LINE; do
      IFS=';' read -r -a BASE_DESC <<< "$LINE"
      echo -e "${GREEN}${BASE_DESC[0]}${NC} ${BASE_DESC[1]}"
      echo -e "  ${BOLD}${BASE_DESC[2]}:${NC} ${BASE_DESC[3]}"
      echo ""
    done
  fi
}

function base {
  echo "Select base image"
}


######################### SCRIPT ARGUMENTS PROCESSING ##########################
exec 3>&1

for i in "$@"; do
  case $i in

    --local|-l ) OPT_LOCAL=true     ;;
    --format=* ) OPT_FORMAT=${i#*=} ;;

  esac
done

case $1 in

  ''             ) interactive ;;

  --version|-v   ) version     ;;
  --help|-h|help ) help        ;;

  update         ) update      ;;
  bases          ) bases       ;;
  base           ) base        ;;
  packages       ) packages    ;;
  add            ) add         ;;
  remove         ) remove      ;;
  set            ) set         ;;
  apply          ) apply       ;;
  discard        ) discard     ;;
  image          ) image       ;;

  *)
    if [ $OPT_LOCAL = true ]; then
      interactive
    else
      echo "$1 is not a valid OS Preparation Utility command!" >&2
      exit 1
    fi

esac
